### PR TITLE
iox-#2329 Add iox::Atomic

### LINKED ISSUE #2329: Add std::atomic abstraction
## Brief feature description

In order to assure that the iceoryx 32-bit support does not break, an `iox::Atomic` needs to be added and used, which ensures that all atomic algorithms are always lock-free.

### PR DESCRIPTION
**Pull request by elBoberido**
## Notes for Reviewer

This PR adds the `iox::Atomic` as drop-in replacement for the std::atomic. The std::atomic will be replaced in a follow up PR.

## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/main/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/main/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/main/doc/website/release-notes/iceoryx-unreleased.md

## Checklist for the PR Reviewer

- [x] Consider a second reviewer for complex new features or larger refactorings
- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] All touched (C/C++) source code files from `iceoryx_hoofs` have been added to `./clang-tidy-diff-scans.txt`
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Relates to #2329

### CODE REVIEW

File: iceoryx_platform/generic/include/iceoryx_platform/atomic.hpp [hunk 1]
```diff
@@ -0,0 +1,518 @@
...
+    static constexpr bool is_always_lock_free = std::atomic<T>::is_always_lock_free;
+
+    /// @brief Constructs a new 'iox::Atomic' with a default value
+    constexpr Atomic() noexcept
+        : m_value{T()}
+    {
+    }
+
+    /// @brief Constructs a new 'iox::Atomic' with the given value
+    explicit constexpr Atomic(T value) noexcept
```
Comment (elBoberido) [inline iceoryx_platform/generic/include/iceoryx_platform/atomic.hpp:59]: This is a deviation from std::atomic but clang-tidy complains when it is not explicit

File: iceoryx_platform/generic/include/iceoryx_platform/atomic.hpp [hunk 2]
```diff
@@ -0,0 +1,518 @@
...
+    /// 'store(value)'
+    // NOLINTNEXTLINE(cppcoreguidelines-c-copy-assignment-signature) This mimics the std::atomic copy assignment operator
+    T operator=(T value) volatile noexcept
+    {
+        m_value = value;
+        return value;
+    }
+
+    /// @brief Atomically loads and returns the stored value. Equivalent to 'load()'
+    explicit operator T() const noexcept
```
Comment (elBoberido) [inline iceoryx_platform/generic/include/iceoryx_platform/atomic.hpp:93]: This is a deviation from std::atomic but clang-tidy complains when it is not explicit

File: iceoryx_platform/generic/include/iceoryx_platform/atomic.hpp [hunk 3]
```diff
@@ -0,0 +1,518 @@
...
+    {
+        return m_value.operator T();
+    }
+
+    /// @brief Atomically loads and returns the stored value. Equivalent to 'load()'
+    explicit operator T() const volatile noexcept
```
Comment (elBoberido) [inline iceoryx_platform/generic/include/iceoryx_platform/atomic.hpp:99]: This is a deviation from std::atomic but clang-tidy complains when it is not explicit

File: iceoryx_platform/test/moduletests/test_platform_atomic.cpp [hunk 4]
```diff
@@ -0,0 +1,761 @@
...
+    EXPECT_THAT(sut_struct.compare_exchange_weak(expected_struct,
+                                                 EXPECTED_INITIAL_STRUCT,
+                                                 std::memory_order::memory_order_release,
+                                                 std::memory_order::memory_order_acquire),
+                Eq(false));
+    EXPECT_THAT(expected_struct, Eq(EXPECTED_NEW_STRUCT));
+    EXPECT_THAT(sut_struct.load(), Eq(EXPECTED_NEW_STRUCT));
+}
+
+TYPED_TEST(Atomic_test, CompareExchangeStronWorks)
```
Comment (elfenpiff) [inline iceoryx_platform/test/moduletests/test_platform_atomic.cpp:384]: Give me a g

**Review by elfenpiff — APPROVED** [review]

**Pull request merged**